server/src/main/java/org/elasticsearch/cluster/NodeConnectionsService.java

```diff
@@ -209,12 +214,36 @@ public void reconnectToNodes(DiscoveryNodes discoveryNodes, Runnable onCompletio
         });
     }
 
-    private class ConnectionTarget {
+    // exposed for testing
+    protected ConnectionTarget connectionTargetForNode(DiscoveryNode node) {
+        synchronized (mutex) {
+            return targetsByNode.get(node);
+        }
+    }
+
+    /**
+     * Time of disconnect in absolute time ({@link ThreadPool#absoluteTimeInMillis()}),
+     * and disconnect-causing exception, if any
+     */
+    record DisconnectionHistory(long disconnectTimeMillis, @Nullable Exception disconnectCause) {
+        public long getDisconnectTimeMillis() {
+            return disconnectTimeMillis;
+        }
+
+        public Exception getDisconnectCause() {
+            return disconnectCause;
+        }
+    }
+
+    protected class ConnectionTarget {
         private final DiscoveryNode discoveryNode;
 
         private final AtomicInteger consecutiveFailureCount = new AtomicInteger();
         private final AtomicReference<Releasable> connectionRef = new AtomicReference<>();
 
+        // access is synchronized by the service mutex
+        protected DisconnectionHistory disconnectionHistory = null;
```

Comment:
nit: suggest marking this nullable (and describing what the null value means)

```suggestion
        @Nullable // if node is connected
        protected DisconnectionHistory disconnectionHistory = null;
```